casanovo/denovo/model.py

```diff
@@ -176,12 +193,592 @@ def forward(
+                            cache_seq_dict[spec_idx].add(pep_str)
+                            smx = self.softmax(scores)[i, :, :]
+
+                            aa_scores = [
+                                smx[idx][self.decoder._aa2idx[aa]].item()
+                                for idx, aa in enumerate(aa_list)
+                            ]
+                            pep_score = np.mean(aa_scores)
+
+                            # Cache peptides with fitting and non-fitting precursor mass separately
+                            if isPrecursorFit:
+                                heappush(
+                                    cache_score_dict[spec_idx][0],
+                                    (pep_score, insert_idx),
+                                )
+                            else:
+                                heappush(
+                                    cache_score_dict[spec_idx][1],
+                                    (pep_score, insert_idx),
+                                )
+
+                        else:
+                            smx = self.softmax(scores)[i, :, :]
+                            aa_scores = [
+                                smx[idx][self.decoder._aa2idx[aa]].item()
+                                for idx, aa in enumerate(aa_list)
+                            ]
+                            new_score = np.mean(aa_scores)
+
+                            # Cache fitting peptide either if non-fitting cached
+                            # or higher confidence than lowest scoring fitting peptide cached
+                            if isPrecursorFit:
+                                # Check if any non-fitting peptide cached
+                                if len(cache_score_dict[spec_idx][1]) > 0:
+                                    (
+                                        pop_pep_score,
+                                        pop_insert_idx,
+                                    ) = heappop(cache_score_dict[spec_idx][1])
+                                    cache_tokens[pop_insert_idx, :] = tokens[
+                                        i, :
+                                    ]
+                                    cache_scores[
+                                        pop_insert_idx, :, :
+                                    ] = scores[i, :, :]
+                                    heappush(
+                                        cache_score_dict[spec_idx][0],
+                                        (new_score, pop_insert_idx),
+                                    )
+
+                                # Check if any fitting peptide cached
+                                elif len(cache_score_dict[spec_idx][0]) > 0:
+                                    (
+                                        pop_pep_score,
+                                        pop_insert_idx,
+                                    ) = heappop(cache_score_dict[spec_idx][0])
```

Comment:
**suggestion:** Can we rewrite the succession of `heappush` and `heappop` more efficiently using `heappushpop`? Right now we're also popping and then pushing the same prediction back on the heap if the score was lower, which is a bit silly.

You can more nicely get the smallest element of a heap using `heap[0]`. That way you could "peek" at the value and only do a `heappushpop` if needed.